### PR TITLE
librdkafka and cppkafka recipes

### PR DESCRIPTION
**Pull request by hmenn**
I couldn't find kafka related recipes in meta-oe and opened this pr.

librdkafka is widely used in new IOT projects

### CODE REVIEW

File: meta-oe/recipes-support/librdkafka/librdkafka_1.6.1.bb
Comment (kraj) [inline meta-oe/recipes-support/librdkafka/librdkafka_1.6.1.bb:15]: why do we need this ?

File: meta-oe/recipes-support/libcppkafka/libcppkafka_git.bb
Comment (kraj) [inline meta-oe/recipes-support/libcppkafka/libcppkafka_git.bb:22]: you will have to `DEPEND` on `chrpath-replacement-native` as well

File: meta-oe/recipes-support/libcppkafka/libcppkafka_git.bb
Comment (kraj) [inline meta-oe/recipes-support/libcppkafka/libcppkafka_git.bb:15]: is this needed ? or can shlibs code automatically detect this dependency and add it to rdeps

**Review by kraj — CHANGES_REQUESTED** [review]

File: meta-oe/recipes-support/libcppkafka/libcppkafka_git.bb
Comment (hmenn) [reply meta-oe/recipes-support/libcppkafka/libcppkafka_git.bb:21]: done

**Review by hmenn — COMMENTED** [review]
done

**Comment by kraj** [comment]
in `master-next` now

**Comment by kraj** [comment]
Fails to configure see

https://errors.yoctoproject.org/Errors/Details/575207/

**Comment by hmenn** [comment]
@kraj Strange. I started a new build on dunfell and couldn't get any error.

**Comment by hmenn** [comment]
@kraj I compiled again on master. Here is the logs. 

```DEBUG: Executing python function extend_recipe_sysroot
NOTE: Direct dependencies are ['virtual:native:/home/hmenn/Workplace/opensource/yocto/poky/meta/recipes-devtools/pseudo/pseudo_git.bb:do_populate_sysroot', 'virtual:native:/home/hmenn/Workplace/opensource/yocto/poky/meta/recipes-devtools/ninja/ninja_1.10.2.bb:do_populate_sysroot', '/home/hmenn/Workplace/opensource/yocto/poky/meta/recipes-devtools/cmake/cmake-native_3.19.5.bb:do_populate_sysroot', '/home/hmenn/Workplace/opensource/yocto/poky/meta/recipes-support/boost/boost_1.75.0.bb:do_populate_sysroot', '/home/hmenn/Workplace/opensource/yocto/meta-openembedded/meta-oe/recipes-extended/librdkafka/librdkafka_1.6.1.bb:do_populate_sysroot', 'virtual:native:/home/hmenn/Workplace/opensource/yocto/poky/meta/recipes-devtools/patch/patch_2.7.6.bb:do_populate_sysroot', '/home/hmenn/Workplace/opensource/yocto/poky/meta/recipes-core/glibc/glibc_2.33.bb:do_populate_sysroot', '/home/hmenn/Workplace/opensource/yocto/poky/meta/recipes-devtools/quilt/quilt-native_0.66.bb:do_populate_sysroot', '/home/hmenn/Workplace/opensource/yocto/poky/meta/recipes-devtools/gcc/gcc-runtime_10.2.bb:do_populate_sysroot', 'virtual:native:/home/hmenn/Workplace/opensource/yocto/poky/meta/recipes-devtools/chrpath/chrpath_0.16.bb:do_populate_sysroot', '/home/hmenn/Workplace/opensource/yocto/poky/meta/recipes-devtools/gcc/gcc-cross_10.2.bb:do_populate_sysroot']
NOTE: Installed into sysroot: []
NOTE: Skipping as already exists in sysroot: ['pseudo-native', 'ninja-native', 'cmake-native', 'boost', 'librdkafka', 'patch-native', 'glibc', 'quilt-native', 'gcc-runtime', 'chrpath-native', 'gcc-cross-x86_64', 're2c-native', 'bzip2-native', 'curl-native', 'xz-native', 'zlib-native', 'ncurses-native', 'zlib', 'bzip2', 'icu', 'python3', 'openssl', 'zstd', 'automake-native', 'attr-native', 'libtool-native', 'autoconf-native', 'linux-libc-headers', 'libgcc', 'binutils-cross-x86_64', 'gmp-native', 'libmpc-native', 'texinfo-dummy-native', 'flex-native', 'mpfr-native', 'gnu-config-native', 'pkgconfig-native', 'openssl-native', 'gettext-minimal-native', 'opkg-utils', 'readline', 'libtirpc', 'libffi', 'libxcrypt', 'sqlite3', 'util-linux', 'libnsl2', 'gdbm', 'xz', 'm4-native', 'unzip-native', 'ncurses', 'libcap-ng', 'bash-completion', 'util-linux-libuuid']
DEBUG: Python function extend_recipe_sysroot finished
DEBUG: Executing shell function do_configure
-- The C compiler identification is GNU 10.2.0
-- The CXX compiler identification is GNU 10.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/hmenn/Workplace/opensource/yocto/build/tmp/work/core2-64-poky-linux/libcppkafka/git-r0/recipe-sysroot-native/usr/bin/x86_64-poky-linux/x86_64-poky-linux-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /home/hmenn/Workplace/opensource/yocto/build/tmp/work/core2-64-poky-linux/libcppkafka/git-r0/recipe-sysroot-native/usr/bin/x86_64-poky-linux/x86_64-poky-linux-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build will generate a shared library. Use CPPKAFKA_BUILD_SHARED=0 to perform a static build
-- Found ZSTD: /home/hmenn/Workplace/opensource/yocto/build/tmp/work/core2-64-poky-linux/libcppkafka/git-r0/recipe-sysroot/usr/lib/libzstd.so
-- RdKafka configuration file found: /home/hmenn/Workplace/opensource/yocto/build/tmp/work/core2-64-poky-linux/libcppkafka/git-r0/recipe-sysroot/usr/lib/cmake/RdKafka/RdKafkaConfig.cmake
-- Disabling tests because submodule Catch2 isn't checked out
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
    LIB_SUFFIX
    PYTHON_EXECUTABLE
    Python3_EXECUTABLE
    Python_EXECUTABLE


-- Build files have been written to: /home/hmenn/Workplace/opensource/yocto/build/tmp/work/core2-64-poky-linux/libcppkafka/git-r0/build
DEBUG: Shell function do_configure finished
DEBUG: Executing python function do_qa_configure
DEBUG: Python function do_qa_configure finished```

This configurations were used.
```
Build Configuration:
BB_VERSION           = "1.49.2"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "universal"
TARGET_SYS           = "x86_64-poky-linux"
MACHINE              = "qemux86-64"
DISTRO               = "poky"
DISTRO_VERSION       = "3.2+snapshot-7d0988966cc1fd9e93549ca2ec21ac3bc765ceb6"
TUNE_FEATURES        = "m64 core2"
TARGET_FPU           = ""
meta                 
meta-poky            
meta-yocto-bsp       = "master:7d0988966cc1fd9e93549ca2ec21ac3bc765ceb6"
meta-oe              
meta-networking      
meta-python          = "master:ba2576ed79a09879f714ccf7f4fa32259c18ab8c"
```

**Comment by kraj** [comment]
the failure is seen with. `master-next` of poky

**Comment by hmenn** [comment]
@kraj Here is the results for master-next branchs.

```
➜  yocto bitbake libcppkafka                       
Loading cache: 100% |##################################################################################################################################################################################################################################| Time: 0:00:01
Loaded 3532 entries from dependency cache.
NOTE: Resolving any missing task queue dependencies

Build Configuration:
BB_VERSION           = "1.49.2"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "universal"
TARGET_SYS           = "x86_64-poky-linux"
MACHINE              = "qemux86-64"
DISTRO               = "poky"
DISTRO_VERSION       = "3.2+snapshot-f5579cd98822c27c3c7414c9e03465f1f54f0637"
TUNE_FEATURES        = "m64 core2"
TARGET_FPU           = ""
meta                 
meta-poky            
meta-yocto-bsp       = "master-next:f5579cd98822c27c3c7414c9e03465f1f54f0637"
meta-oe              
meta-networking      
meta-python          = "master-next:a49892059b8cdfa8e3e5ca8de8e7957355aaa85d"

Initialising tasks: 100% |#############################################################################################################################################################################################################################| Time: 0:00:01
Sstate summary: Wanted 371 Local 0 Network 0 Missed 371 Current 237 (0% match, 38% complete)
Removing 10 stale sstate objects for arch allarch: 100% |##############################################################################################################################################################################################| Time: 0:00:00
Removing 220 stale sstate objects for arch core2-64: 100% |############################################################################################################################################################################################| Time: 0:00:00
Removing 96 stale sstate objects for arch qemux86_64: 100% |###########################################################################################################################################################################################| Time: 0:00:00
Removing 1 stale sstate objects for arch x86_64_ubuntu-18.04: 100% |###################################################################################################################################################################################| Time: 0:00:00
Removing 18 stale sstate objects for arch x86_64: 100% |###############################################################################################################################################################################################| Time: 0:00:00
NOTE: Executing Tasks
NOTE: Tasks Summary: Attempted 1955 tasks of which 1624 didn't need to be rerun and all succeeded.
```

**Comment by kraj** [comment]
I have restaged it in `master-next` and will see if it passes this time.

**Comment by kraj** [comment]
it still shows up, I think the reason is that librdkafka generates cmake with absolute paths into its build directory
and when the component build directory is deleted when using rm_work then this issue appears 

I see 
```
set_target_properties(RdKafka::rdkafka PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "/mnt/b/yoe/master/build/tmp/work/cortexa55-yoe-linux/librdkafka/1.6.1
-r0/recipe-sysroot/usr/include;/mnt/b/yoe/master/build/tmp/work/cortexa55-yoe-linux/librdkafka/1.6.1-r
0/recipe-sysroot/usr/include;/mnt/b/yoe/master/build/tmp/work/cortexa55-yoe-linux/librdkafka/1.6.1-r0/
recipe-sysroot/usr/include;${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "m;ZLIB::ZLIB;/mnt/b/yoe/master/build/tmp/work/cortexa55-yoe-linux/librdkaf
ka/1.6.1-r0/recipe-sysroot/usr/lib/libzstd.so;OpenSSL::SSL;OpenSSL::Crypto;Threads::Threads;dl"
)
```

in `./build/CMakeFiles/Export/lib/cmake/RdKafka/RdKafkaTargets.cmake`

and this .cmake file is exported by librdkafka package

these absolute paths is the problem,  

jenkins builders user rm_work so you can reproducre it like below
add `inherit rm_work` to `librdkafka` recipe and repeat these steps
```
bitbake librdkafka libcppkafka -ccleansstate 
bitbake librdkafka
bitbake libcppkafka
```
and you should see the issue

File: meta-oe/recipes-support/librdkafka/librdkafka/0001_fix_absolute_path_usage.patch
Comment (kraj) [inline meta-oe/recipes-support/librdkafka/librdkafka/0001_fix_absolute_path_usage.patch:5]: please submit this patch upstream as well. If its not worthy of upstream then added the status to patch comment so we know we have to carry it for a reason.

Comment (hmenn) [reply meta-oe/recipes-support/librdkafka/librdkafka/0001_fix_absolute_path_usage.patch:5]: @kraj I've already pushed. Just waiting for response. Thanks 

https://github.com/edenhill/librdkafka/pull/3327

**Comment by kraj** [comment]
still fails :( with latest patch, here are steps.

```diff
diff --git a/meta-oe/recipes-support/librdkafka/librdkafka_1.6.1.bb b/meta-oe/recipes-support/librdkafka/librdkafka_1.6.1.bb
index 21a130e82b..f6be5cf1d5 100644
--- a/meta-oe/recipes-support/librdkafka/librdkafka_1.6.1.bb
+++ b/meta-oe/recipes-support/librdkafka/librdkafka_1.6.1.bb
@@ -19,3 +19,5 @@ inherit cmake
 S = "${WORKDIR}/git"

 FILES_${PN} += "${datadir}"
+
+inherit rm_work
```

```
bitbake -ccleansstate librdkafka libcppkafka
bitbake librdkafka ; bitbake libcppkafka
```

**Comment by hmenn** [comment]
@kraj After lots of debugging I found problems with include path's of external packages. I changed their dependency because we don't them public to use librdkafka unless their FindPackage changed. I'll also push those changes to mainstream. I'm able to build with rm_work now.

For my previous commit, I don't know how build passed with rm_work. I'm really sorry for that mistake.

**Comment by hmenn** [comment]
Hello @kraj, thank you for your support. Now, it's merged with master. What should I do in order to get these in dunfell? Which documentation should I read to learn meta-openembedded merge mechanism?

**Comment by kraj** [comment]
> Hello @kraj, thank you for your support. Now, it's merged with master. What should I do in order to get these in dunfell? Which documentation should I read to learn meta-openembedded merge mechanism?

usually new recipes are not often backported but you can make a case by creating a pull request against dunfell banch and stable maintainer will advice you accordingly.

**Comment by kraj** [comment]
its committed with 3e5209b691c9f4c720335d4f00dc143264c2df96 and 285e612ac9c1ab1fc9f50728cb27b0bd0c40e89c

**Pull request closed without merging**
